ADObjectPermissionEntry: Occasional error creating PSDrive

<!--
    Your feedback and support is greatly appreciated, thanks for contributing!

    ISSUE TITLE:
    Please prefix the issue title with the resource name, e.g.
    'ResourceName: Short description of my issue'

    ISSUE DESCRIPTION (this template):
    Please provide information regarding your issue under each header below.
    Write N/A under any headers that do not apply to your issue, or if the
    information is not available.

    NOTE! Sensitive information should be obfuscated.

    PLEASE KEEP THE HEADERS.

    You may remove this comment block, and the other comment blocks,
    but please keep the headers.
-->
#### Details of the scenario you tried and the problem that is occurring
As part of a full SQL Server AlwaysOn configuration, I am attempting to set the permissions on the AlwaysOn Listener computer object. Occasionally this will fail with the following error:
```
PowerShell DSC resource MSFT_ADObjectPermissionEntry  failed to execute Set-TargetResource functionality with error
message: System.InvalidOperationException: Error creating AD: PS Drive. (ADCOMMON0033) --->
System.Management.Automation.ProviderNotFoundException: Cannot find a provider with the name 'ActiveDirectory'.
   at System.Management.Automation.SessionStateInternal.GetProvider(PSSnapinQualifiedName providerName)
   at System.Management.Automation.SessionStateInternal.GetSingleProvider(String name)
   at Microsoft.PowerShell.Commands.NewPSDriveCommand.ProcessRecord()
   --- End of inner exception stack trace ---
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : Cluster-Node-1
```

I am running the following modules:
ActiveDirectoryDsc: 4.2.0.0
ActiveDirectory: 1.0.0.0 (as ships with Server 2016 build 14393.2551)

This error only occurs with the `ADObjectPermissionEntry` resource, and not everytime I deploy and run the DSC configuration. But once the error occurs, it will always occur for every `ADObjectPermissionEntry` resource, but other resources I am using (`ADManagedServiceAccount`, `ADServicePrincipalName`, and `ADComputer`) never fail with this issue.

#### Verbose logs showing the problem
```
VERBOSE: [Cluster-Node-1]: LCM:  [ Start  Resource ]  [[ADComputer]AgListenerAccount]
VERBOSE: [Cluster-Node-1]: LCM:  [ Start  Test     ]  [[ADComputer]AgListenerAccount]
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Importing the module MSFT_ADComputer in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Importing the module MSFT_ADComputer in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Determining the current state of the computer account 'Listener-Account'. (ADC0006)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Retrieving the information about the computer account 'Listener-Account' from Active Directory. (ADC0002)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] The computer account 'Listener-Account' is present in Active Directory. (ADC0003)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Evaluating the state of the property 'Description'. (ADCOMMON0003)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] The parameter 'Description' is in desired state. (ADCOMMON0004)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] Evaluating the state of the property 'Path'. (ADCOMMON0003)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] The parameter 'Path' is in desired state. (ADCOMMON0004)
VERBOSE: [Cluster-Node-1]:                            [[ADComputer]AgListenerAccount] The computer account 'Listener-Account' is in the desired state. (ADC0011)
VERBOSE: [Cluster-Node-1]: LCM:  [ End    Test     ]  [[ADComputer]AgListenerAccount]  in 0.4290 seconds.
VERBOSE: [Cluster-Node-1]: LCM:  [ Skip   Set      ]  [[ADComputer]AgListenerAccount]
VERBOSE: [Cluster-Node-1]: LCM:  [ End    Resource ]  [[ADComputer]AgListenerAccount]
VERBOSE: [Cluster-Node-1]: LCM:  [ Start  Resource ]  [[ADObjectPermissionEntry]ClusterFullControlToAgListener]
VERBOSE: [Cluster-Node-1]: LCM:  [ Start  Test     ]  [[ADObjectPermissionEntry]ClusterFullControlToAgListener]
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Importing the module MSFT_ADObjectPermissionEntry in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Importing the module MSFT_ADObjectPermissionEntry in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Object permission entry not found on object 'CN=Listener-Account,OU=Servers,DC=example,DC=com'. (OPE0002)
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Object permission entry on object 'CN=Listener-Account,OU=Servers,DC=example,DC=com' is not in the desired state. (OPE0006)
VERBOSE: [Cluster-Node-1]: LCM:  [ End    Test     ]  [[ADObjectPermissionEntry]ClusterFullControlToAgListener]  in 0.4270 seconds.
VERBOSE: [Cluster-Node-1]: LCM:  [ Start  Set      ]  [[ADObjectPermissionEntry]ClusterFullControlToAgListener]
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Importing the module MSFT_ADObjectPermissionEntry in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Importing the module MSFT_ADObjectPermissionEntry in force mode.
VERBOSE: [Cluster-Node-1]:                            [[ADObjectPermissionEntry]ClusterFullControlToAgListener] Creating new AD: PSDrive. (ADCOMMON0032)
VERBOSE: [Cluster-Node-1]: LCM:  [ End    Set      ]  [[ADObjectPermissionEntry]ClusterFullControlToAgListener]  in 0.3820 seconds.
PowerShell DSC resource MSFT_ADObjectPermissionEntry  failed to execute Set-TargetResource functionality with error message: System.InvalidOperationException: Error creating AD: PS Drive. (ADCOMMON0033) --->
System.Management.Automation.ProviderNotFoundException: Cannot find a provider with the name 'ActiveDirectory'.
   at System.Management.Automation.SessionStateInternal.GetProvider(PSSnapinQualifiedName providerName)
   at System.Management.Automation.SessionStateInternal.GetSingleProvider(String name)
   at Microsoft.PowerShell.Commands.NewPSDriveCommand.ProcessRecord()
   --- End of inner exception stack trace ---
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : Cluster-Node-1
```
#### Suggested solution to the issue

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
Relevant section. The full DSC is much greater than this component.
```powershell
$ConfigData = @{
    AllNodes = @(
        @{
            NodeName      = 'Cluster-Node-1'
            DNSDomainName = 'example.com'
            OuPath        = 'OU=Servers,DC=example,DC=com'
            ClusterName   = 'Cluster-Account'
            ListenerName  = 'Listener-Account'
        }
    )
}

Configuration ClusterPermissions {

    Import-DscResource -ModuleName ActiveDirectoryDsc

    WindowsFeature ActiveDirectoryPowerShell {
        Name   = 'RSAT-AD-PowerShell'
        Ensure = 'Present'
    }

    ADComputer AgListenerAccount {
        Ensure               = 'Present'
        DependsOn            = '[WindowsFeature]ActiveDirectoryPowerShell'
        PsDscRunAsCredential = $DomainAdminCredential

        ComputerName      = $Node.ListenerName
        Path              = $Node.OuPath
        Description       = 'SQL Server AlwaysOn listener name account'
    }

    ADObjectPermissionEntry ClusterFullControlToAgListener {
        Ensure                             = 'Present'
        DependsOn                          = '[ADComputer]AgListenerAccount'
        PsDscRunAsCredential               = $DomainAdminCredential

        Path                               = "CN=$($Node.ListenerName),$($Node.OuPath)"
        IdentityReference                  = "$($Node.DNSDomainName)\$($Node.ClusterName)$"
        AccessControlType                  = 'Allow'
        ActiveDirectoryRights              = 'GenericAll'
        ObjectType                         = '00000000-0000-0000-0000-000000000000'
        ActiveDirectorySecurityInheritance = 'None'
        InheritedObjectType                = '00000000-0000-0000-0000-000000000000'
    }
}
```

#### The operating system the target node is running
```
OsName               : Microsoft Windows Server 2016 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsBuildLabEx    : 14393.2551.amd64fre.rs1_release.181004-1309
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

#### Version and build of PowerShell the target node is running
```
Name                           Value
----                           -----
PSVersion                      5.1.14393.2515
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14393.2515
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

#### Version of the DSC module that was used ('dev' if using current dev branch)
4.2.0.0

I suspect the issue is that the AD module is still loaded, and it fails trying to load again. The modules I have loaded just before it tries to load the AD module are:
```

Name                            Version
----                            -------
ActiveDirectory                 1.0.0.0
Microsoft.PowerShell.Management 3.1.0.0
Microsoft.PowerShell.Security   3.0.0.0
Microsoft.PowerShell.Utility    3.1.0.0
Microsoft.WSMan.Management      3.0.0.0
MSFT_ADComputer                 0.0    
MSFT_ADObjectPermissionEntry    0.0    
```

And the PSDrives available are:
```
Name     Provider   
----     --------   
Alias    Alias      
C        FileSystem 
Cert     Certificate
D        FileSystem 
E        FileSystem 
Env      Environment
F        FileSystem 
Function Function   
HKCU     Registry   
HKLM     Registry   
Variable Variable   
WSMan    WSMan      
```
Interestingly, the AD Module is loaded but the AD drive is not available.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,8 @@
   - Resource Integration tests added.
 - Changes to ADKDSKey
   - Added Integration testing ([issue #351](https://github.com/PowerShell/ActiveDirectoryDsc/issues/351))
+- Changes to ADObjectPermissionEntry
+  - Updated Assert-ADPSDrive with PSProvider Checks ([issue #527](https://github.com/PowerShell/ActiveDirectoryDsc/issues/527)).
 
 ## 4.2.0.0
 

--- a/Modules/ActiveDirectoryDsc.Common/ActiveDirectoryDsc.Common.psm1
+++ b/Modules/ActiveDirectoryDsc.Common/ActiveDirectoryDsc.Common.psm1
@@ -1882,23 +1882,34 @@ function Test-DscPropertyState
 
 <#
     .SYNOPSIS
-        Tests if the AD PS Provider has been loaded.
+        Asserts if the AD PS Provider has been installed.
 
-    .OUTPUTS
-        System.Boolean
-
-        Returns $true of the AD PS Provider is loaded, or $false if it is not loaded.
+    .NOTES
+        Attempts to force import the ActiveDirectory module if the AD PS Provider has not been installed
+        and throws an exception if the AD PS Provider cannot be installed.
 #>
 
-function Test-ADPSProvider
+function Assert-ADPSProvider
 {
     [CmdletBinding()]
-    [OutputType([System.Boolean])]
     param()
 
     $activeDirectoryPSProvider = Get-PSProvider -PSProvider 'ActiveDirectory' -ErrorAction SilentlyContinue
 
-    return ($activeDirectoryPSProvider -and $true)
+    if ($null -eq $activeDirectoryPSProvider)
+    {
+        Write-Verbose -Message $script:localizedData.AdPsProviderNotFound -Verbose
+        Import-Module -Name 'ActiveDirectory' -Force
+        try
+        {
+            $activeDirectoryPSProvider = Get-PSProvider -PSProvider 'ActiveDirectory'
+        }
+        catch
+        {
+            $errorMessage = $script:localizedData.AdPsProviderInstallFailureError
+            New-InvalidOperationException -Message $errorMessage -ErrorRecord $_
+        }
+    }
 }
 
 <#
@@ -1909,7 +1920,7 @@ function Test-ADPSProvider
         Specifies the AD path to which the drive is mapped.
 
     .NOTES
-        Throws an exception if the PS Drive cannot be created or the AD PS Provider can't be installed.
+        Throws an exception if the PS Drive cannot be created.
 #>
 function Assert-ADPSDrive
 {
@@ -1923,16 +1934,7 @@ function Assert-ADPSDrive
 
     Assert-Module -ModuleName 'ActiveDirectory'
 
-    if (Test-ADPSProvider)
-    {
-        Write-Verbose -Message $script:localizedData.AdPsProviderNotFound -Verbose
-        Import-Module -Name 'ActiveDirectory' -Force
-        if (Test-ADPSProvider)
-        {
-            $errorMessage = $script:localizedData.AdPsProviderInstallFailureError
-            New-InvalidOperationException -Message $errorMessage -ErrorRecord $_
-        }
-    }
+    Assert-ADPSProvider
 
     $activeDirectoryPSDrive = Get-PSDrive -Name AD -ErrorAction SilentlyContinue
 

--- a/Modules/ActiveDirectoryDsc.Common/ActiveDirectoryDsc.Common.psm1
+++ b/Modules/ActiveDirectoryDsc.Common/ActiveDirectoryDsc.Common.psm1
@@ -1882,13 +1882,34 @@ function Test-DscPropertyState
 
 <#
     .SYNOPSIS
+        Tests if the AD PS Provider has been loaded.
+
+    .OUTPUTS
+        System.Boolean
+
+        Returns $true of the AD PS Provider is loaded, or $false if it is not loaded.
+#>
+
+function Test-ADPSProvider
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
+    param()
+
+    $activeDirectoryPSProvider = Get-PSProvider -PSProvider 'ActiveDirectory' -ErrorAction SilentlyContinue
+
+    return ($activeDirectoryPSProvider -and $true)
+}
+
+<#
+    .SYNOPSIS
         Asserts if the AD PS Drive has been created, and creates one if not.
 
     .PARAMETER Root
         Specifies the AD path to which the drive is mapped.
 
     .NOTES
-        Throws an exception if the PS Drive cannot be created.
+        Throws an exception if the PS Drive cannot be created or the AD PS Provider can't be installed.
 #>
 function Assert-ADPSDrive
 {
@@ -1900,7 +1921,18 @@ function Assert-ADPSDrive
         $Root = '//RootDSE/'
     )
 
-    Assert-Module -ModuleName 'ActiveDirectory' -ImportModule
+    Assert-Module -ModuleName 'ActiveDirectory'
+
+    if (Test-ADPSProvider)
+    {
+        Write-Verbose -Message $script:localizedData.AdPsProviderNotFound -Verbose
+        Import-Module -Name 'ActiveDirectory' -Force
+        if (Test-ADPSProvider)
+        {
+            $errorMessage = $script:localizedData.AdPsProviderInstallFailureError
+            New-InvalidOperationException -Message $errorMessage -ErrorRecord $_
+        }
+    }
 
     $activeDirectoryPSDrive = Get-PSDrive -Name AD -ErrorAction SilentlyContinue
 

--- a/Modules/ActiveDirectoryDsc.Common/en-US/ActiveDirectoryDsc.Common.strings.psd1
+++ b/Modules/ActiveDirectoryDsc.Common/en-US/ActiveDirectoryDsc.Common.strings.psd1
@@ -51,4 +51,6 @@ ConvertFrom-StringData @'
     SearchingForDomainControllerInSite  = Searching for a domain controller in the site '{0}' in the domain '{1}'. (ADCOMMON0053)
     IgnoreCredentialError               = Suppressing the credential error '{0}' with the message '{1}'. (ADCOMMON0054)
     NoObjectFoundInRecycleBin           = Did not find a restorable object in the recycle bin. (ADCOMMON0055)
+    AdPsProviderNotFound                = The Active Directory PS Provider was not found, Forcing import of the ActiveDirectory module. (ADCOMMON0056)
+    AdPsProviderInstallFailureError     = Error installing the Active Directory PS Provider. (ADCOMMON0057)
 '@

--- a/Tests/Unit/ActiveDirectory.Common.Tests.ps1
+++ b/Tests/Unit/ActiveDirectory.Common.Tests.ps1
@@ -2171,69 +2171,158 @@ InModuleScope 'ActiveDirectoryDsc.Common' {
         }
     }
 
-    Describe 'ActiveDirectoryDsc.Common\Assert-ADPSDrive' {
-        Mock -CommandName Assert-Module
+    Describe 'ActiveDirectoryDsc.Common\Assert-ADPSProvider' {
+        BeforeAll {
 
-        Context 'When the AD PS Drive does not exist and the New-PSDrive function is successful' {
-            Mock -CommandName Get-PSDrive -MockWith { $null }
-            Mock -CommandName New-PSDrive
+        }
+
+        Context 'When the AD PS Provider is installed' {
+            BeforeAll {
+                $mockPSProviderResult = @{
+                    Name = 'ActiveDirectory'
+                }
+
+                Mock -CommandName Get-PSProvider `
+                    -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' } `
+                    -MockWith { $mockPSProviderResult }
+                Mock -CommandName Import-Module
+            }
+
+            It 'Should not throw' {
+                { Assert-ADPSProvider -Verbose } | Should -Not -Throw
+            }
+
+            It 'Should call the expected mocks' {
+                Assert-MockCalled -CommandName Get-PSProvider `
+                    -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' } `
+                    -Exactly -Times 1
+                Assert-MockCalled -CommandName Import-Module -Exactly -Times 0
+            }
+        }
+
+        Context 'When the AD PS Provider is not installed' {
+
+            Context 'When the AD PS Provider is successfully installed by Import-Module' {
+                BeforeAll {
+                    $mockPSProviderResult = @{
+                        Name = 'ActiveDirectory'
+                    }
+
+                    Mock -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' }
+                    Mock -CommandName Import-Module
+                    Mock -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -ne 'SilentlyContinue' } `
+                        -MockWith { $mockPSProviderResult }
+                }
+
+                It 'Should not throw' {
+                    { Assert-ADPSProvider } | Should -Not -Throw
+                }
+
+                It 'Should call the expected mocks' {
+                    Assert-MockCalled -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' } `
+                        -Exactly -Times 1
+                    Assert-MockCalled -CommandName Import-Module -Exactly -Times 1
+                    Assert-MockCalled -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -ne 'SilentlyContinue' } `
+                        -Exactly -Times 1
+                }
+            }
+
+            Context 'When the AD PS Provider is not successfully installed by Import-Module' {
+                BeforeAll {
+                    Mock -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' }
+                    Mock -CommandName Import-Module
+                    Mock -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -ne 'SilentlyContinue' } `
+                        -MockWith { Throw 'Error'}
+                }
+
+                It 'Should throw the correct exception' {
+                    { Assert-ADPSProvider } | Should -Throw $script:localizedData.AdPsProviderInstallFailureError
+                }
+
+                It 'Should call the expected mocks' {
+                    Assert-MockCalled -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -eq 'SilentlyContinue' } `
+                        -Exactly -Times 1
+                    Assert-MockCalled -CommandName Import-Module -Exactly -Times 1
+                    Assert-MockCalled -CommandName Get-PSProvider `
+                        -ParameterFilter { $PSBoundParameters['ErrorAction'] -ne 'SilentlyContinue' } `
+                        -Exactly -Times 1
+                }
+            }
+        }
+    }
+
+    Describe 'ActiveDirectoryDsc.Common\Assert-ADPSDrive' {
+        BeforeAll {
+            $defaultPSDriveRoot = '//RootDSE/'
+            Mock -CommandName Assert-Module
+            Mock -CommandName Assert-ADPSProvider
+        }
+
+        Context 'When the AD PS Drive does not exist' {
+            BeforeAll {
+                Mock -CommandName Get-PSDrive
+            }
+
+            Context 'When the New-PSDrive function is successful' {
+                BeforeAll {
+                    Mock -CommandName New-PSDrive
+                }
+
+                It 'Should not throw' {
+                    { Assert-ADPSDrive } | Should -Not -Throw
+                }
+
+                It 'Should call the expected mocks' {
+                    Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1
+                    Assert-MockCalled -CommandName Get-PSDrive -Exactly -Times 1
+                    Assert-MockCalled -CommandName New-PSDrive `
+                        -ParameterFilter { $Root -eq $defaultPSDriveRoot } `
+                        -Exactly -Times 1
+                }
+            }
+
+            Context 'When the New-PSDrive function is not successful' {
+                BeforeAll {
+                    Mock -CommandName New-PSDrive -MockWith { throw }
+                }
+
+                It 'Should throw the correct error' {
+                    { Assert-ADPSDrive } | Should -Throw $script:localizedString.CreatingNewADPSDriveError
+                }
+
+                It 'Should call the expected mocks' {
+                    Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1
+                    Assert-MockCalled -CommandName Get-PSDrive -Exactly -Times 1
+                    Assert-MockCalled -CommandName New-PSDrive `
+                        -ParameterFilter { $Root -eq $defaultPSDriveRoot } `
+                        -Exactly -Times 1
+                }
+            }
+        }
+
+        Context 'When the AD PS Drive already exists' {
+            BeforeAll {
+                Mock -CommandName Get-PSDrive -MockWith { New-MockObject -Type System.Management.Automation.PSDriveInfo }
+                Mock -CommandName New-PSDrive
+            }
 
             It 'Should not throw' {
                 { Assert-ADPSDrive } | Should -Not -Throw
             }
 
-            It 'Should have called Assert-Module' {
-                Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should have called Get-PSDrive only once' {
-                Assert-MockCalled -CommandName Get-PSDrive -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should have called New-PSDrive only once' {
-                Assert-MockCalled -CommandName New-PSDrive -Exactly -Times 1 -Scope Context
-            }
-        }
-
-        Context 'When the AD PS Drive does not exist and the New-PSDrive function is not successful' {
-            Mock -CommandName Get-PSDrive -MockWith { $null }
-            Mock -CommandName New-PSDrive -MockWith { throw }
-
-            It 'Should throw the correct error' {
-                { Assert-ADPSDrive } | Should -Throw $script:localizedString.CreatingNewADPSDriveError
-            }
-
-            It 'Should call Assert-Module' {
-                Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should call Get-PSDrive once' {
-                Assert-MockCalled -CommandName Get-PSDrive -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should call New-PSDrive once' {
-                Assert-MockCalled -CommandName New-PSDrive -Exactly -Times 1 -Scope Context
-            }
-        }
-
-        Context 'When the AD PS Drive already exists' {
-            Mock -CommandName Get-PSDrive -MockWith { New-MockObject -Type System.Management.Automation.PSDriveInfo }
-            Mock -CommandName New-PSDrive
-
-            It 'Should not throw' {
-              { Assert-ADPSDrive } | Should -Not -Throw
-            }
-
-            It 'Should call Assert-Module only once' {
-                Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should call Get-PSDrive only once' {
-                Assert-MockCalled -CommandName Get-PSDrive -Exactly -Times 1 -Scope Context
-            }
-
-            It 'Should not call New-PSDrive' {
-                Assert-MockCalled -CommandName New-PSDrive -Exactly -Times 0 -Scope Context
+            It 'Should call the expected mocks' {
+                Assert-MockCalled -CommandName Assert-Module -Exactly -Times 1
+                Assert-MockCalled -CommandName Get-PSDrive `
+                    -ParameterFilter { $Name -eq 'AD' } `
+                    -Exactly -Times 1
+                Assert-MockCalled -CommandName New-PSDrive -Exactly -Times 0
             }
         }
     }

